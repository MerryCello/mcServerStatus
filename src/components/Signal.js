import { useEffect, useState } from "react"
import bar1 from "../images/1-5_Signal.png"
import bar2 from "../images/2-5_Signal.png"
import bar3 from "../images/3-5_Signal.png"
import bar4 from "../images/4-5_Signal.png"
import fullSignal from "../images/fullSignal.png"
import noSignal from "../images/noSignal.png"
import searchingSignal from "../images/searchingSignal.gif"


const Signal = ({serverStatus, size, style}) => {
   const [sigImg, setSigImg] = useState(searchingSignal)

   useEffect(() => {
      if (serverStatus?.online)
        setSigImg(fullSignal)
      else if (serverStatus?.loading)
         setSigImg(searchingSignal)
      else
         setSigImg(noSignal)
   }, [serverStatus])

   return (
      <div style={style}>
         <img src={sigImg} style={{height: size + "em", width: "auto"}}/>
      </div>
   );
}

export default Signal;