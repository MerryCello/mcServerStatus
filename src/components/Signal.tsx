import React, {CSSProperties, FC, useEffect, useState} from 'react';
import bar1 from '../images/1-5_Signal.png';
import bar2 from '../images/2-5_Signal.png';
import bar3 from '../images/3-5_Signal.png';
import bar4 from '../images/4-5_Signal.png';
import fullSignal from '../images/fullSignal.png';
import noSignal from '../images/noSignal.png';
import searchingSignal from '../images/searchingSignal.gif';

const VERY_POOR_SIGNAL = 8000;
const POOR_SIGNAL = 4000;
const GOOD_SIGNAL = 2000;
const VERY_GOOD_SIGNAL = 1700;
const EXCELLENT_SIGNAL = 0;

type SignalProps = {
  server: {
    hostname: string;
    online?: boolean;
    loading?: boolean;
    pingAvgMs?: number;
  };
  size: number;
  style: CSSProperties;
};

const Signal: FC<SignalProps> = ({server, size, style}) => {
  const [sigImg, setSigImg] = useState(searchingSignal);

  useEffect(() => {
    if (server?.loading) {
      setSigImg(searchingSignal);
    } else if (server?.online && server?.pingAvgMs) {
      let pingAvgMs = server?.pingAvgMs;
      if (pingAvgMs >= EXCELLENT_SIGNAL && pingAvgMs < VERY_GOOD_SIGNAL) {
        setSigImg(fullSignal);
      } else if (pingAvgMs >= VERY_GOOD_SIGNAL && pingAvgMs < GOOD_SIGNAL) {
        setSigImg(bar4);
      } else if (pingAvgMs >= GOOD_SIGNAL && pingAvgMs < POOR_SIGNAL) {
        setSigImg(bar3);
      } else if (pingAvgMs >= POOR_SIGNAL && pingAvgMs < VERY_POOR_SIGNAL) {
        setSigImg(bar2);
      } else if (pingAvgMs >= VERY_POOR_SIGNAL) {
        setSigImg(bar1);
      }
    } else {
      setSigImg(noSignal);
    }
  }, [server]);

  return (
    <div style={style}>
      <img
        src={sigImg}
        style={{height: size + 'em', width: 'auto'}}
        alt='visual of signal strength'
      />
    </div>
  );
};

export default Signal;
