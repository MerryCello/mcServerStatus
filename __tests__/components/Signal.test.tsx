/** @jest-environment jsdom */
import React, { useState } from 'react';
import { render } from '@testing-library/react';
import Signal from '../../src/components/Signal';

const mockSetSigImg = jest.fn();

describe('Signal', () => {
  const defaultProps = {
    server: {
      hostname: 'localhost',
      online: true,
      loading: false,
      pingAvgMs: 0,
    },
    size: 2,
    style: {},
  };

  afterEach(() => {
    mockSetSigImg.mockClear();
  });

  beforeEach(() => {
    (useState as jest.Mock).mockReturnValueOnce(['searchingSignal', mockSetSigImg]);
  })

  it('should render searching signal when loading', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, loading: true }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('searchingSignal');
    expect(container).toMatchSnapshot();
  });

  it('should render no signal when server is offline', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, online: false }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('noSignal');
    expect(container).toMatchSnapshot();
  });

  it('should render full signal when ping is excellent', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, pingAvgMs: 1000 }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('fullSignal');
    expect(container).toMatchSnapshot();
  });

  it('should render bar4 signal when ping is very good', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, pingAvgMs: 1800 }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('bar4');
    expect(container).toMatchSnapshot();
  });

  it('should render bar3 signal when ping is good', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, pingAvgMs: 3000 }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('bar3');
    expect(container).toMatchSnapshot();
  });

  it('should render bar2 signal when ping is poor', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, pingAvgMs: 5000 }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('bar2');
    expect(container).toMatchSnapshot();
  });

  it('should render bar1 signal when ping is very poor', () => {
    const { container } = render(<Signal {...defaultProps} server={{ ...defaultProps.server, pingAvgMs: 9000 }} />);
    expect(mockSetSigImg).toHaveBeenCalledWith('bar1');
    expect(container).toMatchSnapshot();
  });
});