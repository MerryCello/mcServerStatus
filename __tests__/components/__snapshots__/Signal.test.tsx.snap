// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signal should render bar1 signal when ping is very poor 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render bar2 signal when ping is poor 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render bar3 signal when ping is good 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render bar4 signal when ping is very good 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render full signal when ping is excellent 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render no signal when server is offline 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;

exports[`Signal should render searching signal when loading 1`] = `
<div>
  <div>
    <img
      alt="visual of signal strength"
      src="test-file-stub"
      style="height: 2em; width: auto;"
    />
  </div>
</div>
`;
